### PR TITLE
Enable user consent when command requests resources

### LINKED ISSUE #67: Enable user consent to access the local resources/credentials
In the manifest file, command should declare their intention to access the execution context provided by command launcher at runtime. This is particular important when accessing sensitive information like credentials.

During the command execution time, command launcher should request user consent to allow these command access such information. 

Ex. the "test-cmd" requests to access the "LOGIN_TOKEN" and "USERNAME" context.
```
pkgName: test-pkg
version: 1.0.0
cmds:
  - name: test-cmd
    type: executable
    executable: "{{.Root}}/bin/exec.sh"
    context: [ "LOGIN_TOKEN", "USERNAME" ]
```

Example consent message:
```
$ cl test-cmd
test-cmd requests to access following context:
- LOGIN_TOKEN
- USERNAME
Allow access? y/N
```

The user consent should be stored to avoid requesting consent each time.

### PR DESCRIPTION
**Pull request by bhou**
- add new command property: `requestedResources`.
- add new configuration `enable_user_consent`
- add new configuration `user_consent_life`
- refactor the integration tests

close #67

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -409,6 +407,37 @@ func addCommands(groups []command.Command, executables []command.Command) {
...
+			console.Error("Failed to parse arguments: %v", err)
+			// set exit code to 1
+			return envVars, 1, shouldQuit
+		}
+		if h, exist := envTable[fmt.Sprintf("%s_FLAG_HELP", envVarPrefix)]; exist && h == "true" {
+			c.Help()
+			return envVars, 0, shouldQuit
+		}
+		shouldQuit = false
+	} else {
```
Comment (gmonceyron) [inline cmd/root.go:433]: You could avoid this else block by initializing shouldQuit to false

File: test/integration/test-consent.sh [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+echo $RESULT | grep -q "test-password"
+if [ $? -eq 0 ]; then
+  echo "OK"
+else
+  echo "KO - should pass PASSWORD resource to command"
+  exit 1
+fi
+
+echo "> test consent authorized - should not re-request authorization"
```
Comment (gmonceyron) [inline test/integration/test-consent.sh:97]:
```suggestion
echo "> test consent authorized - should not request authorization again"
```

File: test/integration/test-consent.sh [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+$CL_PATH config enable_user_consent true
+RESULT=$(echo 'y' | $CL_PATH bonjour)
+echo $RESULT
+echo $RESULT | grep -q "authorize the access?"
+if [ $? -eq 0 ]; then
+  echo "KO - should NOT re-request authorization"
```
Comment (gmonceyron) [inline test/integration/test-consent.sh:103]:
```suggestion
  echo "KO - should NOT request authorization again"
```

File: test/integration/test-consent.sh [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+echo "> test consent authorized - should request authorization once expired"
+sleep 5
+$CL_PATH config enable_user_consent true
+RESULT=$(echo 'y' | $CL_PATH bonjour)
+echo $RESULT
+echo $RESULT | grep -q "authorize the access?"
+if [ $? -eq 0 ]; then
+  echo "OK"
+else
+  echo "KO - should re-request authorization once expired"
```
Comment (gmonceyron) [inline test/integration/test-consent.sh:118]:
```suggestion
  echo "KO - should request authorization again once expired"
```

File: internal/config/load.go [hunk 5]
```diff
@@ -101,6 +101,9 @@ func setDefaultConfig() {
 
 	viper.SetDefault(CI_ENABLED_KEY, false)
 	viper.SetDefault(PACKAGE_LOCK_FILE_KEY, filepath.Join(appDir, "lock.json"))
+
+	viper.SetDefault(ENABLE_USER_CONSENT_KEY, false)
```
[the remark this answers is not in the document]
Comment (bhou) [reply internal/config/load.go:105]: The default false is to ensure that the current command lines are still working

File: cmd/consent/consent_test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"testing"
+
+	"github.com/criteo/command-launcher/internal/context"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAccessConsents(t *testing.T) {
+	// TODO: we shouldn't let the secret lib depends on the context
```
[the remark this answers is not in the document]
Comment (bhou) [reply cmd/consent/consent_test.go:11]: In today's implementation, the helper function accesses the global variable `context` to retrieve and store secrets. I think it is not a good design. The context variable is only useful in the root command. The helper function (a library) should not rely on the existence of the global variable. 

Another PR will fix the helper function accessing the global variable issue. The comment here is a marker to fix the test.

File: cmd/consent/consent.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		return requests, nil
+	}
+
+	consent, err := getCmdConsents(cmdGroup, cmdName)
+	if err == nil {
+		return consent.Consents, nil
+	}
+
+	if requestConsent(cmdGroup, cmdName, requests) {
+		// yes
```
[the remark this answers is not in the document]
Comment (bhou) [reply cmd/consent/consent.go]: Done in the new commit

File: cmd/consent/consent.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		// expire in seconds for 30 days = 3600 * 24 * 30 = 2592000
```
[the remark this answers is not in the document]
Comment (bhou) [reply cmd/consent/consent.go]: Done in the new commit

File: internal/config/load.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gmonceyron) [reply internal/config/load.go:105]: I think we need to have a look at the existing package to make sure this is taken into account, then moving this back to true in another PR.

**Review by gmonceyron — APPROVED** [review]

**Pull request merged**